NIFI-4185 Add XML Record Reader

JohannesDaniel opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

JohannesDaniel commented:
@pvillard31 here we go :)

markap14 commented:
@JohannesDaniel thanks for the contribution! This is definitely something that I've been wanting to implement for a while but haven't had a chance yet. Will be happy to review.

JohannesDaniel commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/resources/docs/org.apache.nifi.xml.XMLReader/additionalDetails.html:
```diff
@@ -0,0 +1,378 @@
+    </p>
+
+    <code>
+            <pre>
```
done

JohannesDaniel commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLReader.java:
```diff
@@ -0,0 +1,121 @@
+        "XML data, embedded in an enclosing root tag.")
+public class XMLReader extends SchemaRegistryService implements RecordReaderFactory {
+
+    public static final PropertyDescriptor VALIDATE_ROOT_TAG = new PropertyDescriptor.Builder()
```
done. additionally, I added some tests for class XMLReader

JohannesDaniel commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/resources/docs/org.apache.nifi.xml.XMLReader/additionalDetails.html:
```diff
@@ -0,0 +1,378 @@
+
+    <p>
+        For further explanation of the logic of this reader, an example of an array of records shall be demonstrated.
+        The following record contains the field "array_element", which repeatedly occurs. The field contains two
```
done

JohannesDaniel commented:
Hi @pvillard31 

thank you for your comments! I realized all your suggestions. I like your news regarding the performance :-) Which kind of transformation did you test? XML => Record or XML => JSON (e. g. with ConvertRecord)?

For any reason some tests disappeared for a certain commit at my local git (probably, I wanted to reorder the tests,  but deleted them, omg ...). However, I inserted them again (this is why there are many more tests now). 

In addition, I adjusted the definition about how namespaces shall be treated. 

I implemented several tests for XMLReader to verify that the usage of expression language works as expected.

However, I was not able to reproduce your observation regarding the empty record for the header 
```
<?xml version="1.0" encoding="utf-8"?>
```

I implemented the following tests:
```
testSimpleRecordWithHeader()
testSimpleRecordWithHeaderNoValidation()
```

Actually, they work as expected.

JohannesDaniel commented:
Can you maybe post the XML that led to the empty record?

pvillard31 commented:
Hey @JohannesDaniel - just made new tests and can't reproduce what I saw previously... so let's forget about it :) (thanks for the unit tests though!)

I tested the EL support for tag validation, working well, thanks for the addition. I'd just make one suggestion: right now for the record tag, "in the case of a mismatch, the respective record will be skipped". I'd suggest to make this behavior configurable through a parameter in the CS: skip the record (current behavior) or throw an exception for the flow file (as you're doing for the root tag validation).

Regarding the performance test, it was a GFF => ConvertRecord (XML to JSON) => UpdateAttribute. Don't remember the exact numbers but IIRC it was around 1300 flowfile per second with an XML payload of 6KB (single record in my case).

Overall, it's really cool!
I'll try to have a look over the code during the week.

markap14 commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java:
```diff
@@ -0,0 +1,502 @@
+                    xmlEventReader.nextEvent();
+                    if (hasAttributes) {
+                        if (contentFieldName != null) {
+                            recordValues.put(contentFieldName, characters.toString());
```
I think we need to avoid using characters.toString() and use getData() instead

markap14 commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java:
```diff
@@ -0,0 +1,502 @@
+        final Iterator iterator = startElement.getAttributes();
+        while (iterator.hasNext()) {
+            final Attribute attribute = (Attribute) iterator.next();
+            final String attributeName = attribute.getName().toString();
```
Should probably use getName().getLocalPart() here as well.

markap14 commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java:
```diff
@@ -0,0 +1,502 @@
+            if (oldValues instanceof List) {
+                ((List) oldValues).add(fieldValue);
+            } else {
+                values.put(fieldName, new ArrayList<Object>(){{ add(oldValues); add(fieldValue); }});
```
Would avoid the anonymous inner class here too.

markap14 commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java:
```diff
@@ -0,0 +1,502 @@
+            case DATE:
+            case TIME:
+            case TIMESTAMP: {
+                return DataTypeUtils.convertType(character.toString(), dataType, LAZY_DATE_FORMAT, LAZY_TIME_FORMAT, LAZY_TIMESTAMP_FORMAT, fieldName);
```
Should use getData() here too

markap14 commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java:
```diff
@@ -0,0 +1,502 @@
+        return null;
+    }
+
+    private Object parseCharacterForType(Characters character, String fieldName, DataType dataType) {
```
Probably can create a single parseStringDataForType(String data, String fieldName, DataType dataType) method and call that from both parseAttributeForType and parseCharacterForType, rather than having to repeat the large case statement.

markap14 commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java:
```diff
@@ -0,0 +1,502 @@
+                        if (contentFieldName != null) {
+                            recordValues.put(contentFieldName, characters.toString());
+                        } else {
+                            logger.debug("Found content for field that has to be parsed as record but property \"Field Name for Content\" is not set. " +
```
I wonder if it makes sense in this case to use the name of the field  as the 'contentFieldName'. For example, if we had the following XML snippet:
```
<person>
  <location gurantee="NONE">1.469, -3.875</location>
</person>
```
Then if the Content Field Name were null, we would just use the name of the field (in this case 'location') so that a flattened representation of the record would look like:
```
person.location.guarantee = NONE
person.location.location = 1.469, -3.875
```
Thoughts?

JohannesDaniel replied:
nice idea

JohannesDaniel replied:
But what if the xml has the following structure?

```
<person>
  <location location="NONE">1.469, -3.875</location>
</person>
```

then the will replace the attribute as long as location ist not defined as an array...

markap14 commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLReader.java:
```diff
@@ -0,0 +1,133 @@
+import java.util.Map;
+
+@Tags({"xml", "record", "reader", "parser"})
+@CapabilityDescription("Reads XML content and creates Record objects. Records are expected in the second level of " +
```
More specifically, I think that if the following were the content of a FlowFile:
```
<person>
  <name>John Doe</name>
  <id>123</id>
  <dob>01/01/2017</dob>
</person>
```
Then I would expect to have this parse as a single Record that would match this schema:
```
{
  "name": "person", "namespace": "nifi",
  "type": "record",
  "fields": [
    { "name": "name", "type": "string" },
    { "name": "id", "type": "int" },
    { "name": "dob", "type": "date" }
  ]
}
```
Additionally, I would expect to be able to set a property that indicates that the outer-most XML element is simply a wrapper. If that property were set to "true", then I would expect to use that exact same schema to parse the following XML:
```
<people>
  <person>
    <name>John Doe</name>
    <id>123</id>
    <dob>01/01/2017</dob>
  </person>
  <person>
    <name>Jane Doe</name>
    <id>124</id>
    <dob>01/01/2016</dob>
  </person>
  <person>
    <name>Jake Doe</name>
    <id>125</id>
    <dob>01/01/2015</dob>
  </person>
</people>
```
In this case, the 'people' element is just a wrapper and could just as easily be an element named 'root' or 'foo' or 'bar'.

JohannesDaniel replied:
when I started implementing this reader, I was wondering, how the reader knows whether to parse wrapped records or a single record. unfortunately we dont have an unambiguous indicator like we have for json: [ vs. { 
I considered to make it configurable with EL whether the reader shall expect a single record or an array of records. what do you think?

markap14 replied:
Yes, exactly. I would use a property to convey this. I would be okay with allowing Expression Language to be used, or just allowing for a 'true'/'false' without Expression Language (I think in most cases, you'll want one or the other, not dependent upon each individual FlowFile). But if you think EL is important then I won't argue that point :)
One other option, which we do in a few different processors, would be to offer a third option that looks at a well-known attribute. So you could choose 'true' (treat outer element as a wrapper), 'false' (treat each flowfile as a single record), or 'use xml.stream attribute', and when that is selected, the 'xml.stream' attribute would be looked at to determine how to handle it - a value of 'true' would mean it's a stream of multiple records, 'false' would mean it's only 1 record, missing or any other value would throw an Exception. I don't have  strong preference one way or another how this should be handled, but wanted to present options that we typically use.

JohannesDaniel commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLReader.java:
```diff
@@ -0,0 +1,133 @@
+            .required(false)
+            .build();
+
+    public static final PropertyDescriptor VALIDATE_RECORD_TAG = new PropertyDescriptor.Builder()
```
my original intention actually was to enable users to parse recordsets like this
```
<root>
  <record>...
  <record>...
  <non-record>...

JohannesDaniel replied:
(non-record shall be skipped)

JohannesDaniel commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLReader.java:
```diff
@@ -0,0 +1,133 @@
+    public static final PropertyDescriptor ATTRIBUTE_PREFIX = new PropertyDescriptor.Builder()
+            .name("attribute_prefix")
+            .displayName("Attribute Prefix")
+            .description("If this property is set, the name of attributes will be appended by a prefix when they are added to a record.")
```
ok

JohannesDaniel commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java:
```diff
@@ -0,0 +1,502 @@
+            final XMLInputFactory xmlInputFactory = XMLInputFactory.newInstance();
+
+            // Avoid namespace replacements
+            xmlInputFactory.setProperty(XMLInputFactory.IS_NAMESPACE_AWARE, false);
```
ok, I will activate namespaces and implement some tests for this.

JohannesDaniel commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java:
```diff
@@ -0,0 +1,502 @@
+                        if (oldValues instanceof List) {
+                            ((List) oldValues).add(newValue);
+                        } else {
+                            return new ArrayList<Object>(){{ add(oldValues); add(newValue); }};
```
why?

JohannesDaniel commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java:
```diff
@@ -0,0 +1,502 @@
+                return parseRecord(startElement, childSchema, true, true);
+            }
+            case MAP: {
+                logger.warn("Type map is not supported by this record reader. Field will be skipped.");
```
ok, I ignored this so far as I had no clear idea how a map could look like in xml

JohannesDaniel commented:
@markap14 thank you for the comprehensive review. I will start refactoring the implementations with respect to the improvements that are clear.

JohannesDaniel commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java:
```diff
@@ -0,0 +1,502 @@
+                        return new ArrayList<Object>(){{ add(newValue); }};
+                    }
+                } else {
+                    return null;
```
This already works. If null is returned, the outer loop doesn't do anything, so the field containing the list remains. However, I changed this to make the logic more obvious.

JohannesDaniel commented:
@markap14 @pvillard31 
- I refactored some code as the cases (coerce==true && drop==false) and (coerce==false && drop==true) in some cases showed an unexpected behavior
- Data like <tag>content<subtag>content</subtag>content</tag> now can be parsed
- Maps (e. g. <map_field><key1>value1</key1><key2>value2</key2></map_field>) are now supported
- The reader is now able to parse single records (e. g. <record><field1/><field2/></record>) as well as arrays of records (e. g. <root><record/><record/></root>). I added a property to make it configurable whether the reader shall expect a single record or an array. One question: As there are only two options for this, I defined AllowableValues for this property. Despite that, I think it would be reasonable to enable EL for this property. But how can this be realized?
- I removed the root validation, but remained the check for record tag names in order to support processing data like this <root><record/><record/><other/></root> (tag "other" will be ignored if check for record tag name is activated)

markap14 commented:
@JohannesDaniel thanks for the update! I commented above re: the use of Expression Language in the property descriptor.
I do still feel like the check for 'record tag names' is unnecessary, as the reader should not be responsible for filtering the data but rather just for reading it. There already exist mechanisms for filtering the data (You could use PartitionRecord + RouteOnAttribute, ValidateRecord, or QueryRecord just off the top of my head to achieve this). Additionally, we have the Schema for the Record Reader. So if the element name matches the top-level Schema name (or one of them, if the top-level field is a UNION/CHOICE element), then we could use that. So, with your example above, if you only want to read the `<record/><record/>` part, your schema should indicate that the top-level field name is `record`. In that case, it should filter out the `other` record. Does that make sense?

JohannesDaniel commented:
@markap14 thank you for the response. 

I will simply remove that record tag validation as there are indeed many ways to do that before the data is processed by this reader. 

There is one little corner case, I need to discuss: 
Assuming we have the following data:
```
<record><map_field><key1>value1</key1><key2>...
```
If the reader is used with (coerce==true), the field "map_field" can be parsed by defining a map in the schema. The embedded key fields do not have to be defined, its values only have to be of the defined type for the map. 

If the reader is used with (coerce==false && dropUnknown==true), the reader will parse all fields that exist in the schema ignoring its type. However, the data above will not be parsable even if the map exists in the schema. In this case, the reader identifies "map_field" as a field that exists in the schema, but the reader is not aware that it is of type map. Therefore, the reader will not parse the embedded key fields, as they don't exist in the schema. The field "map_field" will be classified as an empty field and not added to the record.

Furthermore, even if the reader is used with (coerce==false && dropUnknown==true), it will be type-aware to some extent. The reader first checks, whether fields exist in the schema. If that is the case, the reader additionally will check whether they are of type record (or of type array embedding records, respectively). If that is also the case, the reader will retrieve the subschema in order to be enabled to check whether subtags of the current tag are known.

JohannesDaniel commented:
@markap14 
- Added EL for record format property
- Removed record tag validation

tballison commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/xml/XMLRecordReader.java:
```diff
@@ -84,6 +84,10 @@ public XMLRecordReader(InputStream in, RecordSchema schema, String rootName, Str
             final XMLInputFactory xmlInputFactory = XMLInputFactory.newInstance();
+
+            // Avoid namespace replacements
+            xmlInputFactory.setProperty(XMLInputFactory.IS_NAMESPACE_AWARE, false);
```
Might want to avoid XEE vulnerability via improved configuration of XMLInputFactory?
https://www.owasp.org/index.php/XML_External_Entity_(XXE)_Prevention_Cheat_Sheet#XMLInputFactory_.28a_StAX_parser.29

JohannesDaniel replied:
@tballison Thank you for the advice! I refactored this in a way that only the local part is considered. 
:)

JohannesDaniel commented:
@markap14 thanks for the help with the patch file. I am fine with it :)

If you have not planned to do that yourself, I would start implementing a XMLWriter as soon as this has been merged. Or should we first discuss your plan with the Record attributes you mentioned above?

markap14 commented:
@JohannesDaniel that would be great! I think the Record attribute stuff will significantly improve how we are able to handle XML-based records. But I think the approach that you've taken here is a great first step and will be very beneficial to the community. I'd recommend going ahead with the XML Writer and then we can work in the attribute-related stuff later.
